Documentation/DesignDocs/corewcfqueue-design.md

```diff
@@ -0,0 +1,308 @@
+}
+
+```
+- For StartAsync
+ - First we create a list of tasks`StartFetchingMessage` for each `QueueTransportContext` (populated during Init).
+ - `QueueTransportPump` from `QueueTransportContext` is used by each `StartFetchingMessage` to call the `StartPumpAsync` method.
+ - If the `QueueTransportPump` is based on a push message mechanism, it awaits until a message is received and invokes service as soon as the message is received.
+ - If the `QueueTransportPump` is using `DefaultQueueTransportPump`, it creates multiple concurrent tasks to invoke `FetchAndProcessAsync` method based on concurrency level.
+ -The `FetchAndProcessAsync` method of QueueTransport (concrete transport) calls `ReceiveQueueMessageContextAsync` to retrieve the message and pass the handle to the middleware for processing. 
+- StopAsync cancels all tasks and, as a result, the fetching of messages is stopped.
+#### QueueFetchMessage
+- `QueueFetchMessage` is the first middleware which is invoked as soon as a message is fetched from the queue.
+- The `InvokeAsync` method reads messages from pipereader and encodes them into a concrete `Message` objet and puts them back to `QueueMessageContext` and invokes the next delegate , which is `QueueProcessMessage`.
+#### QueueProcessMessage
+- Invokes the `DispatchAsync` method of the channel dispatcher by passing `QueueMessageContext` object and ending the middleware pipeline.
+- As a part of `DispatchAsync`, the service operation is called, processing of messages is done, eventually `QueueMessageContext` (inherits `RequestContext`) `ReplyAsync` method is called.
+- The `ReplyAsync` mechanism is used to inform the concrete transport that a message processing has finished with a status of success or failure by invoking an Action `DispatchResultHandler`. 
+- By calling the `DispatchResultHandler` each queue providers which support timeouts or requires notification to prevent reprocessing a message can take appropriate actions. 
+- The reply message sent via `DispatchResultHandler` could be a fault message, which will signal a failure to process the message, and can be used to provide an error message back to the queue provider.
+
+# Workflow Diagram
+![CoreWCF Generic Queue](/Documentation/DesignDocs/corewcf_queue.png?raw=true)
+
+
+ - During the server registrations, the QueuePollingService initializes the service dispatcher and gets the concrete transport from each queue layer transport.
+ - When the ASP.NET Core WebHost starts, it calls IHostedService. StartAsync on the QueuePollingService, creates a list of tasks (StartFetchingMessage) for each transport and runs them in a loop.
+ - Each task again creates a list of subtasks based on the QueueTransport (concurrent property) to retrieve message context(QueueMessageContext) (until server stops)
+ - The fetched QueueMessageContext is passed to the middleware. The first middleware layer reads the message bytes from a PipeReader and decodes them to a Message object.
```

Comment:
Name the concurrent property with ClassName.PropertyName and backtick quote.. Remove (until server stops) from parenthesis and instead add a comma just before that phrase. Also worth mentioning until Webhost calls QueuePollingService.StopAsync as the mechanism that it will be stopped.